code4romania/monitorizare-vot
# Feature/issue 317 data export improvement

### What does it fix?


Closes #317 



### Changes
- **VoteMonitor.Api.DataExport**
   - Change query execution from entity framework to dapper (https://stackexchange.github.io/Dapper/)
     - why dapper ? Because dapper has focus on performance: https://github.com/StackExchange/Dapper#performance. in addition it has been created by Stackoverflow.
   - on query export change return type form `List<ExportModel>` to `IEnumerable<ExportModel>` to avoid materialization operation (dapper query result return an IEnumerable itself)
   - Add CSV Exporter using same library of `VoteMonitor.Api.County` to export a generic `IEnumerable`
- **VotingIrregularities.Domain.Seed**
   - add migration to create missing index `IX_Answer_IdObserver_CountyCode_PollingStationNumber_LastModified`

### How has it been tested?

1. Added fake answer data to my test database (1474317 rows)
![image](https://user-images.githubusercontent.com/9533250/96364136-747ba380-1139-11eb-9a1f-ead3a153300e.png)
1.1 All `LastModified` are grater than default value `2019-11-08`
![image](https://user-images.githubusercontent.com/9533250/96364226-04b9e880-113a-11eb-88f9-335b7fb8da5b.png)

2. Execute multiple times export api `/api​/v1​/export​/all` (via swagger) without providing any parameters, in order to get filter all data in my database
3. Check the result
3.1 application insights timing: around 25s
![image](https://user-images.githubusercontent.com/9533250/96364246-21eeb700-113a-11eb-81d7-c253fdc573b3.png)
3.2 exported file size ~260 Mb, 1474318 rows
![image](https://user-images.githubusercontent.com/9533250/96364276-4185df80-113a-11eb-8386-7477221c7851.png)
